src/cosmos/CosmosEngine.ts

```diff
@@ -218,44 +194,143 @@ export class CosmosEngine extends CurrencyEngine<
+    let txCountTotal
+    let txCount = 0
+    let earlyExit = false
+    try {
+      do {
+        const txRes = await tmClient.txSearch({
+          ...txSearchParams,
+          page: ++page
+        })
+        const { totalCount, txs } = txRes
+        if (txCountTotal == null) txCountTotal = totalCount
+        txCount = txCount + txs.length
+
+        for (const tx of txs) {
+          const txidHex = toHex(tx.hash).toUpperCase()
+          if (newestTxid == null) {
+            newestTxid = txidHex
+          }
 
-    const isSend = from === this.walletInfo.keys.bech32Address
-    const ourReceiveAddresses: string[] = []
+          if (txidHex === this.otherData[queryString]) {
+            earlyExit = true
+            break
+          }
 
-    const currencyCode = value.denom.toUpperCase()
-    let nativeAmount = value.amount
+          // The bank module emits 'transfer' events with sender, recipient, and coins
+          const transferEvents: TransferEvent[] = []
+          const parsedLogs = parseRawLog(tx.result.log)
+          parsedLogs.forEach(log => {
+            log.events.forEach(event => {
+              const transferEvent = asMaybe(asTransfer)(event)
+              if (transferEvent == null) return []
+              const [recipient, sender, amount] = transferEvent.attributes
+
+              const coins = parseCoins(amount.value)
+              return coins.forEach(coin => {
+                transferEvents.push({
+                  sender: sender.value,
+                  recipient: recipient.value,
+                  coin
+                })
+              })
+            })
+          })
+          if (transferEvents.length === 0) continue
+
+          const block = await client.getBlock(tx.height)
+          const date = toSeconds(
+            fromRfc3339WithNanoseconds(block.header.time)
+          ).seconds
+          transferEvents.forEach(event => {
+            this.processCosmosTransaction(txidHex, date, event, tx)
+          })
+        }
+      } while (txCountTotal != null && txCountTotal > txCount && !earlyExit)
+    } catch (e) {
+      this.log.warn('queryTransactions error:', e)
+      throw e
+    }
 
+    return newestTxid
+  }
+
+  processCosmosTransaction(
+    txidHex: string,
+    date: number,
+    event: TransferEvent,
+    tx: TxResponse
+  ): void {
+    const { height, tx: txRaw } = tx
+    const signedTx = base16.stringify(txRaw)
+    const {
+      authInfo: { fee },
+      body: { memo }
+    } = decodeTxRaw(txRaw)
+
+    let networkFeeCoin = this.networkInfo.defaultTransactionFee
+    if (fee != null) {
+      const { amount } = fee
+      networkFeeCoin = safeAddCoins([networkFeeCoin, ...amount])
+    }
+    const networkFee = networkFeeCoin.amount
+
+    const { coin, recipient, sender } = event
+
+    if (
+      this.walletInfo.keys.bech32Address !== recipient &&
+      this.walletInfo.keys.bech32Address !== sender &&
+      this.currencyInfo.currencyCode !== coin.denom.toUpperCase()
```

Comment:
if currencycode does match, but the address doesn't match as sender or recipient, why would we include this tx?